File: app/domain/button.py
from uuid import uuid4
from app.domain.icon import Icon
from app.domain.macro import Macro


class Button:
    
    uuid: str
    name: str
    icon: Icon
    macro: Macro
    
    def __init__(self, name: str, icon: Icon, macro: Macro, uuid):
        self.name = name
        self.icon = icon
        self.macro = macro
        if uuid is None:
            self.uuid = str(uuid4())
        else:
            self.uuid = uuid
    
    def to_dict(self):
        return {
            "uuid": self.uuid,
            "name": self.name,
            "icon": self.icon.to_dict(),
            "macro": self.macro.to_dict()
        }
        
    @classmethod
    def from_dict(cls, data):
        return cls(
            name=data["name"],
            icon=Icon.from_dict(data["icon"]),
            macro=Macro.from_dict(data["macro"]),
            uuid=data["uuid"]
        )

File: kahldeck-dashboard/src/domain/Button.ts
import { v4 as uuidv4 } from 'uuid';
import { Icon } from './Icon';
import { Macro } from './Macro';

export class Button {
    readonly uuid: string;
    name: string;
    icon: Icon;
    macro: Macro;

    constructor(name: string, icon: Icon, macro: Macro, uuid?: string) {
        this.name = name;
        this.icon = icon;
        this.macro = macro;
        this.uuid = uuid || uuidv4();
    }

    toJSON(): Record<string, unknown> {
        return {
            uuid: this.uuid,
            name: this.name,
            icon: this.icon.toJSON(),
            macro: this.macro.toJSON()
        };
    }

    static fromJSON(data: Record<string, never>): Button {
        return new Button(
            data.name,
            Icon.fromJSON(data.icon),
            Macro.fromJSON(data.macro),
            data.uuid
        );
    }
}